MapTransactionContextTest:  member softKill and Mama,   HazelcastSerializationException + IegalStateException: Nested....

3.3-RC4-SNAPSHOT git log -1 --format="%H" 8c8fc7ec51a485a1b6a208b87065f60da652d48c

2 ec2 instances
2 members
2 clients
3m duration

txnCon.class=com.hazelcast.stabilizer.tests.map.MapTransactionContextTest
txnCon.basename=txnCon
txnCon.threadCount=3
txnCon.keyCount=1000

Same test as for https://github.com/hazelcast/hazelcast/issues/3195,  just Diffrent exceptions

3 worker client instances Hit this Serialization Exception

```
Failure[
   message='Worked ran into an unhandled exception'
   type='Worker exception'
   agentAddress=10.183.40.66
   time=Thu Aug 07 14:04:53 UTC 2014
   workerAddress=client:10.183.40.66
   workerId=worker-10.183.40.66-2-client
   test=TestCase{
          id=txnCon
        , class=com.hazelcast.stabilizer.tests.map.MapTransactionContextTest
        , basename=txnCon
        , keyCount=1000
        , threadCount=3
    }
   cause=com.hazelcast.nio.serialization.HazelcastSerializationException: java.io.NotSerializableException: com.hazelcast.instance.MemberImpl
    at com.hazelcast.nio.serialization.SerializationServiceImpl.handleException(SerializationServiceImpl.java:354)
    at com.hazelcast.nio.serialization.SerializationServiceImpl.toData(SerializationServiceImpl.java:228)
    at com.hazelcast.nio.serialization.SerializationServiceImpl.toData(SerializationServiceImpl.java:200)
    at com.hazelcast.client.ClientEngineImpl.sendResponse(ClientEngineImpl.java:153)
    at com.hazelcast.client.ClientEndpoint.sendResponse(ClientEndpoint.java:254)
    at com.hazelcast.client.CallableClientRequest.process(CallableClientRequest.java:31)
    at com.hazelcast.client.ClientEngineImpl$ClientPacketProcessor.processRequest(ClientEngineImpl.java:391)
    at com.hazelcast.client.ClientEngineImpl$ClientPacketProcessor.run(ClientEngineImpl.java:334)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:745)
    at com.hazelcast.util.executor.PoolExecutorThreadFactory$ManagedThread.run(PoolExecutorThreadFactory.java:59)
    at ------ End remote and begin local stack-trace ------.(Unknown Source)
    at com.hazelcast.client.spi.impl.ClientCallFuture.resolveResponse(ClientCallFuture.java:194)
    at com.hazelcast.client.spi.impl.ClientCallFuture.get(ClientCallFuture.java:136)
    at com.hazelcast.client.spi.impl.ClientCallFuture.get(ClientCallFuture.java:113)
    at com.hazelcast.client.txn.TransactionProxy.invoke(TransactionProxy.java:200)
    at com.hazelcast.client.txn.TransactionProxy.commit(TransactionProxy.java:131)
    at com.hazelcast.client.txn.TransactionContextProxy.commitTransaction(TransactionContextProxy.java:84)
    at com.hazelcast.stabilizer.tests.map.MapTransactionContextTest$Worker.run(MapTransactionContextTest.java:77)
    at java.lang.Thread.run(Thread.java:745)
    at com.hazelcast.stabilizer.tests.utils.ThreadSpawner$DefaultThread.run(ThreadSpawner.java:88)
Caused by: java.io.NotSerializableException: com.hazelcast.instance.MemberImpl
    at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1183)
    at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1547)
    at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1508)
    at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1431)
    at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1177)
    at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1547)
    at java.io.ObjectOutputStream.defaultWriteObject(ObjectOutputStream.java:440)
    at java.lang.Throwable.writeObject(Throwable.java:985)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:988)
    at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1495)
    at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1431)
    at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1177)
    at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:347)
    at com.hazelcast.nio.serialization.DefaultSerializers$ObjectSerializer.write(DefaultSerializers.java:216)
    at com.hazelcast.nio.serialization.StreamSerializerAdapter.write(StreamSerializerAdapter.java:52)
    at com.hazelcast.nio.serialization.SerializationServiceImpl.toData(SerializationServiceImpl.java:219)
    at com.hazelcast.nio.serialization.SerializationServiceImpl.toData(SerializationServiceImpl.java:200)
    at com.hazelcast.client.ClientEngineImpl.sendResponse(ClientEngineImpl.java:153)
    at com.hazelcast.client.ClientEndpoint.sendResponse(ClientEndpoint.java:254)
    at com.hazelcast.client.CallableClientRequest.process(CallableClientRequest.java:31)
    at com.hazelcast.client.ClientEngineImpl$ClientPacketProcessor.processRequest(ClientEngineImpl.java:391)
    at com.hazelcast.client.ClientEngineImpl$ClientPacketProcessor.run(ClientEngineImpl.java:334)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:745)
    at com.hazelcast.util.executor.PoolExecutorThreadFactory$ManagedThread.run(PoolExecutorThreadFactory.java:59)
]
```

1 client worker thread hit, the Nested transactions are not allowed

```
Failure[
   message='Worked ran into an unhandled exception'
   type='Worker exception'
   agentAddress=10.237.178.186
   time=Thu Aug 07 14:04:49 UTC 2014
   workerAddress=client:10.237.178.186
   workerId=worker-10.237.178.186-2-client
   test=TestCase{
          id=txnCon
        , class=com.hazelcast.stabilizer.tests.map.MapTransactionContextTest
        , basename=txnCon
        , keyCount=1000
        , threadCount=3
    }
   cause=java.lang.IllegalStateException: Nested transactions are not allowed!
    at com.hazelcast.client.txn.TransactionProxy.begin(TransactionProxy.java:93)
    at com.hazelcast.client.txn.TransactionContextProxy.beginTransaction(TransactionContextProxy.java:80)
    at com.hazelcast.stabilizer.tests.map.MapTransactionContextTest$Worker.run(MapTransactionContextTest.java:68)
    at java.lang.Thread.run(Thread.java:745)
    at com.hazelcast.stabilizer.tests.utils.ThreadSpawner$DefaultThread.run(ThreadSpawner.java:88)
]
```

worker standard out,  we can see 2 client worker thread did not hit exceptions,
"TxnCounter{committed=11381, rolled=0} from 2 workers"

```
find . -name "out.log" | xargs -I % sh -c 'echo %; cat %; echo"" '
./workers/1407420188737/worker-10.183.40.66-3-server/out.log

./workers/1407420188737/worker-10.183.40.66-2-client/out.log
txnCon: com.hazelcast.spi.exception.TargetDisconnectedException: Target[Address[10.237.178.186]:5701] disconnected.
txnCon: com.hazelcast.spi.exception.TargetDisconnectedException: Target[Address[10.237.178.186]:5701] disconnected.

./workers/1407420188737/worker-10.237.178.186-2-client/out.log
txnCon: com.hazelcast.core.HazelcastInstanceNotActiveException: Hazelcast instance is not active!

./workers/1407420188737/worker-10.183.40.66-1-server/out.log
txnCon: TxnCounter{committed=11381, rolled=0} from 2 workers
txnCon: received increments from 2 workers

./workers/1407420188737/worker-10.237.178.186-1-server/out.log
```

--- a/hazelcast/src/main/java/com/hazelcast/core/MemberLeftException.java
+++ b/hazelcast/src/main/java/com/hazelcast/core/MemberLeftException.java
@@ -17,26 +17,27 @@
 package com.hazelcast.core;
 
 import com.hazelcast.instance.MemberImpl;
-import com.hazelcast.nio.ObjectDataInput;
-import com.hazelcast.nio.ObjectDataOutput;
-import com.hazelcast.nio.serialization.DataSerializable;
+import com.hazelcast.nio.Address;
 import com.hazelcast.spi.exception.RetryableException;
 
 import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.net.InetSocketAddress;
 import java.util.concurrent.ExecutionException;
 
 /**
  * Thrown when a member left during an invocation or execution.
  */
-@edu.umd.cs.findbugs.annotations.SuppressWarnings("SE_BAD_FIELD")
-public class MemberLeftException extends ExecutionException implements DataSerializable, RetryableException {
+public class MemberLeftException extends ExecutionException implements RetryableException {
 
-    private Member member;
+    private transient Member member;
 
     public MemberLeftException() {
     }
 
     public MemberLeftException(Member member) {
+        super(member + " has left cluster!");
         this.member = member;
     }
 
@@ -48,16 +49,40 @@ public class MemberLeftException extends ExecutionException implements DataSeria
         return member;
     }
 
-    public String getMessage() {
-        return member + " has left cluster!";
+    private void writeObject(ObjectOutputStream out) throws IOException {
+        out.defaultWriteObject();
+
+        out.writeUTF(member.getUuid());
+
+        boolean isImpl = member instanceof MemberImpl;
+        out.writeBoolean(isImpl);
+
+        if (isImpl) {
+            MemberImpl memberImpl = (MemberImpl) member;
+            Address address = memberImpl.getAddress();
+            String host  = address.getHost();
+            int port = address.getPort();
+            out.writeUTF(host);
+            out.writeInt(port);
+        } else {
+            InetSocketAddress socketAddress = member.getSocketAddress();
+            out.writeObject(socketAddress);
+        }
     }
 
-    public void writeData(ObjectDataOutput out) throws IOException {
-        member.writeData(out);
-    }
+    private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException {
+        in.defaultReadObject();
 
-    public void readData(ObjectDataInput in) throws IOException {
-        member = new MemberImpl();
-        member.readData(in);
+        String uuid = in.readUTF();
+
+        boolean isImpl = in.readBoolean();
+        if (isImpl) {
+            String host = in.readUTF();
+            int port = in.readInt();
+            member = new MemberImpl(new Address(host, port), false, uuid, null);
+        } else {
+            InetSocketAddress socketAddress = (InetSocketAddress) in.readObject();
+            member = new MemberImpl(new Address(socketAddress), false, uuid, null);
+        }
     }
 }

--- a/hazelcast/src/test/java/com/hazelcast/instance/SimpleMemberImpl.java
+++ b/hazelcast/src/test/java/com/hazelcast/instance/SimpleMemberImpl.java
@@ -1,0 +1,164 @@
+/*
+ * Copyright (c) 2008-2014, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.instance;
+
+import com.hazelcast.core.Member;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.Map;
+
+/**
+ * @author mdogan 25/08/14
+ */
+public class SimpleMemberImpl implements Member {
+
+    private String uuid;
+    private InetSocketAddress address;
+
+    public SimpleMemberImpl() {
+    }
+
+    public SimpleMemberImpl(String uuid, InetSocketAddress address) {
+        this.uuid = uuid;
+        this.address = address;
+    }
+
+    @Override
+    public boolean localMember() {
+        return false;
+    }
+
+    @Override
+    public InetSocketAddress getInetSocketAddress() {
+        return getSocketAddress();
+    }
+
+    @Override
+    public InetSocketAddress getSocketAddress() {
+        return address;
+    }
+
+    @Override
+    public String getUuid() {
+        return uuid;
+    }
+
+    @Override
+    public Map<String, Object> getAttributes() {
+        return null;
+    }
+
+    @Override
+    public String getStringAttribute(String key) {
+        return null;
+    }
+
+    @Override
+    public void setStringAttribute(String key, String value) {
+
+    }
+
+    @Override
+    public Boolean getBooleanAttribute(String key) {
+        return null;
+    }
+
+    @Override
+    public void setBooleanAttribute(String key, boolean value) {
+
+    }
+
+    @Override
+    public Byte getByteAttribute(String key) {
+        return null;
+    }
+
+    @Override
+    public void setByteAttribute(String key, byte value) {
+
+    }
+
+    @Override
+    public Short getShortAttribute(String key) {
+        return null;
+    }
+
+    @Override
+    public void setShortAttribute(String key, short value) {
+
+    }
+
+    @Override
+    public Integer getIntAttribute(String key) {
+        return null;
+    }
+
+    @Override
+    public void setIntAttribute(String key, int value) {
+
+    }
+
+    @Override
+    public Long getLongAttribute(String key) {
+        return null;
+    }
+
+    @Override
+    public void setLongAttribute(String key, long value) {
+
+    }
+
+    @Override
+    public Float getFloatAttribute(String key) {
+        return null;
+    }
+
+    @Override
+    public void setFloatAttribute(String key, float value) {
+
+    }
+
+    @Override
+    public Double getDoubleAttribute(String key) {
+        return null;
+    }
+
+    @Override
+    public void setDoubleAttribute(String key, double value) {
+
+    }
+
+    @Override
+    public void removeAttribute(String key) {
+
+    }
+
+    @Override
+    public void writeData(ObjectDataOutput out) throws IOException {
+        out.writeUTF(uuid);
+        out.writeObject(address);
+    }
+
+    @Override
+    public void readData(ObjectDataInput in) throws IOException {
+        uuid = in.readUTF();
+        address = in.readObject();
+    }
+}

--- a/hazelcast/src/test/java/com/hazelcast/nio/serialization/SerializationTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/nio/serialization/SerializationTest.java
@@ -23,6 +23,11 @@ import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.core.IExecutorService;
 import com.hazelcast.executor.impl.CancellationOperation;
 import com.hazelcast.nio.IOUtil;
+import com.hazelcast.core.Member;
+import com.hazelcast.core.MemberLeftException;
+import com.hazelcast.instance.MemberImpl;
+import com.hazelcast.instance.SimpleMemberImpl;
+import com.hazelcast.nio.Address;
 import com.hazelcast.nio.ObjectDataInput;
 import com.hazelcast.nio.ObjectDataOutput;
 import com.hazelcast.spi.OperationAccessor;
@@ -36,8 +41,13 @@ import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
 import java.io.Serializable;
+import java.net.InetSocketAddress;
 import java.util.ArrayList;
 import java.util.LinkedList;
 import java.util.Random;
@@ -318,5 +328,45 @@ public class SerializationTest
                 return Foo.this;
             }
         }
+    }
+
+    @Test
+    public void testMemberLeftException_usingMemberImpl() throws IOException, ClassNotFoundException {
+        String uuid = UuidUtil.buildRandomUuidString();
+        String host = "127.0.0.1";
+        int port = 5000;
+
+        Member member = new MemberImpl(new Address(host, port), false, uuid, null);
+
+        testMemberLeftException(uuid, host, port, member);
+    }
+
+    @Test
+    public void testMemberLeftException_usingSimpleMember() throws IOException, ClassNotFoundException {
+        String uuid = UuidUtil.buildRandomUuidString();
+        String host = "127.0.0.1";
+        int port = 5000;
+
+        Member member = new SimpleMemberImpl(uuid, new InetSocketAddress(host, port));
+        testMemberLeftException(uuid, host, port, member);
+    }
+
+    private void testMemberLeftException(String uuid, String host, int port, Member member)
+            throws IOException, ClassNotFoundException {
+
+        MemberLeftException exception = new MemberLeftException(member);
+
+        ByteArrayOutputStream bout = new ByteArrayOutputStream();
+        ObjectOutputStream out = new ObjectOutputStream(bout);
+        out.writeObject(exception);
+
+        ByteArrayInputStream bin = new ByteArrayInputStream(bout.toByteArray());
+        ObjectInputStream in = new ObjectInputStream(bin);
+        MemberLeftException exception2 = (MemberLeftException) in.readObject();
+        MemberImpl member2 = (MemberImpl) exception2.getMember();
+
+        assertEquals(uuid, member2.getUuid());
+        assertEquals(host, member2.getAddress().getHost());
+        assertEquals(port, member2.getAddress().getPort());
     }
 }

